### PR TITLE
✨ Bootstrap code with kubebuilder and set up initial CRDs

### PR DESCRIPTION
**Pull request by embik**
## Summary

This adds an initial scaffold via kubebuilder. It's big as it stands, so I'm not adding anything more.

Core focus of the review (i.e. the only things I changed) should be the `apis/` directory. I've written some simple documentation for how objects relate to each other as well.

## Related issue(s)

Fixes #

## Release Notes

```release-note
Add initial API definitions
```

### CODE REVIEW

File: cmd/main.go [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+	utilruntime.Must(operatorkcpiov1alpha1.AddToScheme(scheme))
+	// +kubebuilder:scaffold:scheme
+}
+
+func main() {
+	var metricsAddr string
+	var enableLeaderElection bool
+	var probeAddr string
+	var secureMetrics bool
+	var enableHTTP2 bool
```
Comment (mjudeikis) [inline cmd/main.go:57]: Can we squash:
```
var metricsAddr  probeAddr enableHTTP2 string 
var enableLeaderElection probeAddr bool
```

File: internal/controller/suite_test.go [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+var ctx context.Context
+var cancel context.CancelFunc
+
+func TestControllers(t *testing.T) {
+	RegisterFailHandler(Fail)
+
+	RunSpecs(t, "Controller Suite")
+}
+
+var _ = BeforeSuite(func() {
```
Comment (mjudeikis) [inline internal/controller/suite_test.go:55]: Is ginko something we really want to go with? :D I would prefer something more like kcp `framework`. Wondering what others thing?

**Review by mjudeikis — APPROVED** [review]
In general as skeleton to start with - I think its good to go in and we iterate. Its not like it set in stone. If other are ok, we can move on?

File: internal/controller/suite_test.go (hunk 2 above)
Comment (embik) [reply internal/controller/suite_test.go:55]: This is bootstrapped by kubebuilder. I'm open to changing it once we implement real tests, I'd just leave it in for now.

File: Makefile [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+KUSTOMIZE ?= $(LOCALBIN)/kustomize
+CONTROLLER_GEN ?= $(LOCALBIN)/controller-gen
+ENVTEST ?= $(LOCALBIN)/setup-envtest
+GOLANGCI_LINT = $(LOCALBIN)/golangci-lint
+
+## Tool Versions
+KUSTOMIZE_VERSION ?= v5.4.3
+CONTROLLER_TOOLS_VERSION ?= v0.16.1
+ENVTEST_VERSION ?= release-0.19
+GOLANGCI_LINT_VERSION ?= v1.59.1
```
[the remark this answers is not in the document]
Comment (embik) [reply Makefile:174]: No reason, just how kubebuilder structured this file. Will move.

File: Makefile [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+.PHONY: modules
+modules: ## Run go mod tidy to ensure modules are up to date
+	hack/update-go-modules.sh
+
+$(OPENSHIFT_GOIMPORTS):
+	GOBIN=$(TOOLS_GOBIN_DIR) $(GO_INSTALL) github.com/openshift-eng/openshift-goimports $(OPENSHIFT_GOIMPORTS_BIN) $(OPENSHIFT_GOIMPORTS_VER)
+
+.PHONY: imports
+imports: $(OPENSHIFT_GOIMPORTS)
+	$(OPENSHIFT_GOIMPORTS) -m github.com/kcp-dev/kcp
```
[the remark this answers is not in the document]
Comment (embik) [reply Makefile]: Nope.

**Pull request merged**